### PR TITLE
DAOS-7673 Add storage tiering to daos server yaml and management tools

### PR DESCRIPTION
**Pull request by llasek-intc**
Wide reaching change that adds support for configuration of multiple
storage tiers as specified in the server YAML file. This change aims to
reach functional parity with the master branch whilst building a
framework to use a user-specified number of storage tiers.

### CODE REVIEW

File: src/control/server/storage/class.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	msgBdevBadSize = "backfile_size should be greater than 0"
+)
+
+// bdev describes parameters and behaviors for a particular bdev class.
+type bdev struct {
+	templ   string
+	isEmpty func(*BdevConfig) string                // check no elements
+	isValid func(*BdevConfig) string                // check valid elements
+	init    func(logging.Logger, *BdevConfig) error // prerequisite actions
+}
```
Comment (mjmac) [inline src/control/server/storage/class.go:60]: @tanabarr: Looks like we've got some conflicting approaches to solving this problem that need to be resolved. I'm still digesting this PR but wanted to make you aware of this right away.

File: src/control/server/storage/config.go [hunk 2]
```diff
@@ -313,3 +298,11 @@ func (bc *BdevConfig) Validate(class Class) error {
...
 	return nil
 }
+
+type StorageConfig struct {
+	Tiers      Configs `yaml:"storage"`
+	TiersNum   int     `yaml:"-"` // @todo_llasek: cmdLongFlag:"--nvme_tiers" cmdShortFlag:"-T"`
+	ConfigPath string  `yaml:"-" cmdLongFlag:"--nvme" cmdShortFlag:"-n"`
+	MemSize    int     `yaml:"-" cmdLongFlag:"--mem_size,nonzero" cmdShortFlag:"-r,nonzero"`
+	VosEnv     string  `yaml:"-" cmdEnv:"VOS_BDEV_CLASS"`
+}
```
Comment (mjmac) [inline src/control/server/storage/config.go]: I don't understand this change. Why do we now have `storage.Config` and `storage.StorageConfig`? The idea behind my changes was that a given engine would have N `storage.Config` entries, each representing a single tier.

File: src/control/go.sum [hunk 3]
```diff
@@ -523,6 +526,7 @@ google.golang.org/protobuf v1.25.0/go.mod h1:9JNX74DMeImyA3h4bdi1ymwjUzf21/xIlba
 google.golang.org/protobuf v1.26.0-rc.1/go.mod h1:jlhhOSvTdKEhbULTjvd4ARK9grFBp09yW+WbY/TyQbw=
 google.golang.org/protobuf v1.26.0 h1:bxAC2xTBsZGibn2RTntX0oH50xLsqy1OxA9tTL3p/lk=
 google.golang.org/protobuf v1.26.0/go.mod h1:9q0QmTI4eRPtz6boOQmLYwt+qCgq0jsYwAQnmE0givc=
+gopkg.in/alecthomas/kingpin.v2 v2.2.6 h1:jMFz6MfLP0/4fUyZle81rXUoxOBFi19VUFKVDOQfozc=
```
Comment (mjmac) [inline src/control/go.sum:529]: What is bringing these new dependencies in? Generally speaking, we try to be conservative about adding new dependencies because it increases the release burden and size of the `src/control/vendor` directory.

File: doc/admin/pool_operations.md [hunk 4]
```diff
@@ -12,35 +12,30 @@ manage storage pools from the command line.
 
 **To create a pool:**
 ```bash
-$ dmg pool create --size=NTB
+$ dmg pool create --total-size=NTB
```
Comment (mjmac) [inline doc/admin/pool_operations.md:15]: `total-size` is redundant. Please just use `size`, as this is the parameter name that was chosen for the simplified pool create workflow.

File: doc/admin/pool_operations.md [hunk 5]
```diff
@@ -12,35 +12,30 @@ manage storage pools from the command line.
...
-      -s, --scm-size=  Per-server SCM allocation for DAOS pool (manual)
-      -n, --nvme-size= Per-server NVMe allocation for DAOS pool (manual)
-      -r, --ranks=     Storage server unique identifiers (ranks) for DAOS pool
-      -S, --sys=       DAOS system that pool is to be a part of (default: daos_server)
+      -g, --group=      DAOS pool to be owned by given group, format name@domain
+      -u, --user=       DAOS pool to be owned by given user, format name@domain
+      -p, --name=       Unique name for pool (set as label)
+      -a, --acl-file=   Access Control List file path for DAOS pool
+      -z, --total-size= Total size of DAOS pool (auto)
+      -t, --tier-ratio= Percentage of storage tiers for pool storage (auto) (default: 6,94)
```
Comment (mjmac) [inline doc/admin/pool_operations.md:34]: I recommend rephrasing this to something like, "Distribution of pool storage allocation over storage tiers". I also recommend removing the default value example because it won't make sense when there are > 2 tiers defined.

File: doc/admin/pool_operations.md [hunk 6]
```diff
@@ -12,35 +12,30 @@ manage storage pools from the command line.
...
 [create command options]
-      -g, --group=     DAOS pool to be owned by given group, format name@domain
-      -u, --user=      DAOS pool to be owned by given user, format name@domain
-      -p, --name=      Unique name for pool (set as label)
-      -a, --acl-file=  Access Control List file path for DAOS pool
-      -z, --size=      Total size of DAOS pool (auto)
-      -t, --scm-ratio= Percentage of SCM:NVMe for pool storage (auto) (default: 6)
-      -k, --nranks=    Number of ranks to use (auto) (default: all)
-      -v, --nsvc=      Number of pool service replicas (default: 3)
-      -s, --scm-size=  Per-server SCM allocation for DAOS pool (manual)
```
Comment (mjmac) [inline doc/admin/pool_operations.md:40]: I am generally in favor of removing these parameters because it simplifies this effort, but be advised that there are still a lot of uses of them in the automated tests and user scripts. We may want to keep the parameters for now but emit a warning that they are incompatible with storage tiering and will be removed.

File: src/control/server/storage/class.go (hunk 1 above)
Comment (llasek-intc) [reply src/control/server/storage/class.go:60]: I did move the file from bdev to storage pkg to avoid circular dependency between packages. Bdev does include storage, so couldn't have included bdev from storage.
I don't like it but that enables me to make progress. Can refactor later.

File: src/control/server/mgmt_pool.go [hunk 7]
```diff
@@ -26,7 +26,8 @@ import (
 const (
 	// DefaultPoolScmRatio defines the default SCM:NVMe ratio for
 	// requests that do not specify one.
-	DefaultPoolScmRatio = 0.06
+	DefaultPoolScmRatio  = 0.06
+	DefaultPoolNvmeRatio = 0.94
```
Comment (mjmac) [inline src/control/server/mgmt_pool.go:30]: Hmm. I understand that this was probably used for the sake of expediency in getting this work integrated, but it seems like it would be better to just make the code future proof now. As we discussed, I think a reasonable default here for the Bdev tiers would be `(1 - DefaultPoolScmRatio) / len(req.GetTierratio()))`, i.e. allocate the remaining non-SCM portion of the storage evenly among the Bdev tiers.

File: src/control/server/mgmt_pool.go [hunk 8]
```diff
@@ -101,41 +102,45 @@ func (svc *mgmtSvc) calculateCreateStorage(req *mgmtpb.PoolCreateReq) error {
 	if len(req.GetRanks()) == 0 {
 		return errors.New("zero ranks in calculateCreateStorage()")
 	}
-	if req.GetScmratio() == 0 {
-		req.Scmratio = DefaultPoolScmRatio
+	if len(req.GetTierratio()) == 0 {
+		req.Tierratio = make([]float64, 2)
+		req.Tierratio[0] = DefaultPoolScmRatio
+		req.Tierratio[1] = DefaultPoolNvmeRatio
```
Comment (mjmac) [inline src/control/server/mgmt_pool.go:108]: Definitely wary of using these hard-coded array lengths and indexes. If you want to do it this way in order to just get it working first, with the understanding that it will be improved later, I'm OK with that. We shouldn't release with the code like this though, it's too fragile.

File: src/control/server/storage/config.go (hunk 2 above)
Comment (llasek-intc) [reply src/control/server/storage/config.go]: These are global arguments for the engine.
Note the engine (or rather SPDK) does take a single config file describing all devices.
The VOS env is global anwyay. IMO that can be completely removed but trying to limit the necessary changes for 2.0.
The MemSize is for in-memory blobs. This isn't a production feature, nor did it even work a couple of weeks ago. Leaving as a global value, applicable to all tiers.

File: src/control/go.sum (hunk 3 above)
Comment (llasek-intc) [reply src/control/go.sum:529]: I am a complete new to GO, not sure how to check. This is what I got after compile.
Probably not really needed to be pushed but it is under 'src'.

File: doc/admin/pool_operations.md (hunk 4 above)
Comment (llasek-intc) [reply doc/admin/pool_operations.md:15]: Okay.

File: doc/admin/pool_operations.md (hunk 5 above)
Comment (llasek-intc) [reply doc/admin/pool_operations.md:34]: Okay.

File: src/control/server/mgmt_pool.go (hunk 7 above)
Comment (llasek-intc) [reply src/control/server/mgmt_pool.go:30]: I don't think this is reasonable, i.e. the tiers will differ in size. Optane capacity will be likely much lower than e.g. a QLC NAND.

Comment (mjmac) [reply src/control/server/mgmt_pool.go:30]: Ok, fair enough. My point was more that the code here should set some kind of default values and not assume a fixed number of tiers. I am fine with adding support for improved tier-aware defaults in follow-on work.

File: src/control/go.sum (hunk 3 above)
Comment (mjmac) [reply src/control/go.sum:529]: The tooling will update go.mod and go.sum if you import new packages somewhere. Probably you have these elsewhere in your project, as I don't see any evidence of their use in this PR. You can just revert the changes to those two files in your next commit.

File: src/control/server/mgmt_pool.go (hunk 8 above)
Comment (llasek-intc) [reply src/control/server/mgmt_pool.go:108]: In this release the engine will understand only two tiers: SCM + NVMe. We did not specify how to generate tier ratios so the translation from default values to ratios will be hardcoded. For this case the result is an array with two entries. If specified differently, this will have to be changed anyway.
Note this will be a valid input in the future, while the tool allows the user to pass an arbitrary number of ratios - in that case the code in question does not execute.
I'm sorry, I don't see how the highlighted code would be too fragile.

However, I do see an issue with the code below and handling of arbitrary number of tiers. Changing that.

File: src/control/server/storage/config.go (hunk 2 above)
Comment (mjmac) [reply src/control/server/storage/config.go]: OK, I see where you're going with this now. Please take a look at this PR which should merge cleanly on top of your current PR to improve some naming/clarity: https://github.com/daos-stack/daos/pull/5781

File: doc/admin/pool_operations.md (hunk 6 above)
Comment (llasek-intc) [reply doc/admin/pool_operations.md:40]: Alright, bringing nvme-size and scm-size back.

File: doc/man/man8/dmg.8.orig [hunk 9]
```diff
@@ -0,0 +1,654 @@
+<<<<<<< HEAD
```
Comment (johannlombardi) [inline doc/man/man8/dmg.8.orig:1]: file should be git rm?

File: src/control/cmd/daos_admin/handler.go [hunk 10]
```diff
@@ -11,6 +11,7 @@ import (
 
 	"github.com/daos-stack/daos/src/control/logging"
 	"github.com/daos-stack/daos/src/control/pbin"
+	"github.com/daos-stack/daos/src/control/server/storage"
```
Comment (johannlombardi) [inline src/control/cmd/daos_admin/handler.go:14]: dunno much about go imports, but isn't that a superset of the 2 imports below that should be removed?

**Review by johannlombardi — CHANGES_REQUESTED** [review]
I have only reviewed the tool and config file changes (not the go code). Looks good to me.
That being said, the dmg.8.orig would need to be removed from the PR before landing.

File: doc/man/man8/dmg.8.orig (hunk 9 above)
Comment (llasek-intc) [reply doc/man/man8/dmg.8.orig:1]: Oops, sorry. That was a temporary file, will remove.

File: src/control/cmd/daos_admin/handler.go (hunk 10 above)
Comment (llasek-intc) [reply src/control/cmd/daos_admin/handler.go:14]: All are actually needed - the file uses both scm and bdev storage providers from packages below.
Btw. I'm not an expert in GO either, but noticed the GOlang plugin in vscode does automatically take care of adding/cleaning up pkg list, so it should be all fine.

Comment (tanabarr) [reply src/control/cmd/daos_admin/handler.go:14]: so we have a custom preference which I don't think is handled by vscode (last time I checked), we prefer groupings by stdlib/vendor/local i.e. fmt in first group, protobuf in second group then daos/common in the third group, let me know if that doesn't make sense

Comment (tanabarr) [reply src/control/cmd/daos_admin/handler.go:14]: @johannlombardi it's not actually a superset, the child packages have their own namespaces (storage package is distinct from bdev and scm), it is a bit counterintuitive

Comment (llasek-intc) [reply src/control/cmd/daos_admin/handler.go:14]: Is there a tool for that?

**Review by tanabarr — DISMISSED** [review]
go changes look good to me

File: src/control/cmd/daos_admin/handler.go (hunk 10 above)
Comment (mjmac) [reply src/control/cmd/daos_admin/handler.go:14]: No, imports are done on a package level. Each directory is its own package rather than the root of a hierarchy.

File: src/engine/init.c [hunk 11]
```diff
@@ -85,6 +85,8 @@ hwloc_obj_t		numa_obj;
 int			dss_num_cores_numa_node;
 /** Module facility bitmask */
 static uint64_t		dss_mod_facs;
+/** Number of storage tiers: 2 for SCM and NVMe */
+int dss_storage_tiers = 2;
```
Comment (NiuYawei) [inline src/engine/init.c:89]: I'm not quite sure how this "dss_storage_tiers" will be used by io engine, and I didn't see where it's used in current patch.

Could you elaborate how it will be used?

Comment (llasek-intc) [reply src/engine/init.c:89]: Right, it's not being used atm - only a preparation for further work, otherwise it may be removed.
This was carried over from my fork w/ tiering implementation in the engine. The value will be passed to bio_xsctxt_alloc(), in order to initialize the blobstore ctx for all tiers.

The changes in question are available for review (look for 'dss_nvme_tiers'): https://github.com/llasek-intc/daos/tree/llasek/tiering

**Comment by llasek-intc** [comment]
> I have only reviewed the tool and config file changes (not the go code). Looks good to me.
> That being said, the dmg.8.orig would need to be removed from the PR before landing.

Done.

File: src/control/server/storage/bdev/provider.go [hunk 12]
```diff
@@ -207,12 +108,12 @@ func forwardScan(req ScanRequest, cache *ScanResponse, scan scanFwdFn) (msg stri
...
 
-// Scan attempts to perform a scan to discover NVMe components in the
+// storage.BdevScan attempts to perform a scan to discover NVMe components in the
 // system. Results will be cached at the provider and returned if
 // "NoCache" is set to "false" in the request. Returned results will be
 // filtered by request "DeviceList" and empty filter implies allowing all.
-func (p *Provider) Scan(req ScanRequest) (resp *ScanResponse, err error) {
-	if p.shouldForward(req) {
+func (p *Provider) Scan(req storage.BdevScanRequest) (resp *storage.BdevScanResponse, err error) {
+	/*if p.shouldForward(req) {
```
Comment (mjmac) [inline src/control/server/storage/bdev/provider.go:116]: Should probably just go ahead and remove these commented-out blocks. I think I left them in for reference/scaffolding while this was in progress.

**Review by mjmac — COMMENTED** [review]
Generally looks good, just a few things to clean up. For the remaining test/code that is commented out, I assume the plan is to fix that as part of re-enabling VMD?

File: src/engine/init.c (hunk 11 above)
Comment (NiuYawei) [reply src/engine/init.c:89]: Ok, thanks.

File: src/control/cmd/dmg/utils.go [hunk 13]
```diff
@@ -92,3 +94,22 @@ func errIncompatFlags(key string, incompat ...string) error {
...
 	return errors.Errorf("%s with --%s", base, strings.Join(incompat, " or --"))
 }
+
+func parseUint64Array(in string) (out []uint64, err error) {
+	arr, err := csv.NewReader(strings.NewReader(in)).Read()
+	if arr == nil || err != nil {
+		out = nil
+		return
+	}
+	out = make([]uint64, 0)
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/dmg/utils.go:104]: Thanks. Will do.

File: src/control/cmd/dmg/utils.go [hunk 14]
```diff
@@ -92,3 +94,22 @@ func errIncompatFlags(key string, incompat ...string) error {
 
 	return errors.Errorf("%s with --%s", base, strings.Join(incompat, " or --"))
 }
+
+func parseUint64Array(in string) (out []uint64, err error) {
+	arr, err := csv.NewReader(strings.NewReader(in)).Read()
+	if arr == nil || err != nil {
+		out = nil
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/dmg/utils.go:101]: Thanks. Will simplify.

File: src/control/server/ctl_firmware_test.go [hunk 15]
```diff
@@ -23,6 +20,8 @@ import (
 	"github.com/daos-stack/daos/src/control/server/storage/bdev"
 	"github.com/daos-stack/daos/src/control/server/storage/scm"
 	"github.com/daos-stack/daos/src/control/system"
+	"github.com/google/go-cmp/cmp"
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/server/ctl_firmware_test.go:23]: Will sort it all out.

File: src/control/server/storage/bdev/provider.go (hunk 12 above)
Comment (llasek-intc) [reply src/control/server/storage/bdev/provider.go:116]: Alright.

**Review by mjmac — DISMISSED** [review]

**Review by NiuYawei — APPROVED** [review]
The C changes looks good to me.

File: src/control/server/storage/bdev/provider.go [hunk 16]
```diff
@@ -139,16 +124,6 @@ func (p *Provider) Scan(req storage.BdevScanRequest) (resp *storage.BdevScanResp
 // Prepare attempts to perform all actions necessary to make NVMe
 // components available for use by DAOS.
 func (p *Provider) Prepare(req storage.BdevPrepareRequest) (*storage.BdevPrepareResponse, error) {
-	/*if p.shouldForward(req) {
```
Comment (tanabarr) [inline src/control/server/storage/bdev/provider.go]: I probably missed an earlier discussion but why was this removed? It doesn't seem like it would interfere with your tiering related changes (setting a flag on the request seems orthogonal), I would have preferred to Mark the commented block with a TODO DAOS-XXX comment. This is not a blocking remark.

**Review by tanabarr — APPROVED** [review]

**Comment by tanabarr** [comment]
has the failing workflow build step been identified?

**Comment by llasek-intc** [comment]
> has the failing workflow build step been identified?

@tanabarr : this is internal to the CI and unrelated to the patch:
```
The command '/bin/sh -c dnf -y install deltarpm &&     dnf -y upgrade &&     dnf -y install         boost-python36-devel         clang-analyzer         cmake         CUnit-devel         doxygen         e2fsprogs         file         flex         fuse3-devel         gcc         gcc-c++         git         golang         graphviz         hwloc-devel         ipmctl         java-1.8.0-openjdk         json-c-devel         lcov         libaio-devel         libcmocka-devel         libevent-devel         libipmctl-devel         libiscsi-devel         libtool         libtool-ltdl-devel         libunwind-devel         libuuid-devel         libyaml-devel         Lmod         lz4-devel         make         man         maven         nasm         ndctl         numactl         numactl-devel         openmpi3-devel         openssl-devel         pandoc         patch         patchelf         pciutils         python36-Cython         python36-devel         python36-distro         python36-junit_xml         python36-numpy         python36-paramiko         python36-pylint         python36-requests         python36-tabulate         python36-pyxattr         python36-PyYAML         python36-scons         sg3_utils         sudo         valgrind-devel         yasm &&     dnf clean all' returned a non-zero code: 141
Error: Process completed with exit code 141.
```

Btw. the blue ocean view does report success for all steps.

File: src/control/server/storage/bdev/provider.go (hunk 16 above)
Comment (mjmac) [reply src/control/server/storage/bdev/provider.go]: Because the bdev/scm providers are only called from daos_admin now.

**Comment by tanabarr** [comment]
> > has the failing workflow build step been identified?
> 
> @tanabarr : this is internal to the CI and unrelated to the patch:
> 
> ```
> The command '/bin/sh -c dnf -y install deltarpm &&     dnf -y upgrade &&     dnf -y install         boost-python36-devel         clang-analyzer         cmake         CUnit-devel         doxygen         e2fsprogs         file         flex         fuse3-devel         gcc         gcc-c++         git         golang         graphviz         hwloc-devel         ipmctl         java-1.8.0-openjdk         json-c-devel         lcov         libaio-devel         libcmocka-devel         libevent-devel         libipmctl-devel         libiscsi-devel         libtool         libtool-ltdl-devel         libunwind-devel         libuuid-devel         libyaml-devel         Lmod         lz4-devel         make         man         maven         nasm         ndctl         numactl         numactl-devel         openmpi3-devel         openssl-devel         pandoc         patch         patchelf         pciutils         python36-Cython         python36-devel         python36-distro         python36-junit_xml         python36-numpy         python36-paramiko         python36-pylint         python36-requests         python36-tabulate         python36-pyxattr         python36-PyYAML         python36-scons         sg3_utils         sudo         valgrind-devel         yasm &&     dnf clean all' returned a non-zero code: 141
> Error: Process completed with exit code 141.
> ```
> 
> Btw. the blue ocean view does report success for all steps.
> Another thing is an automated check:
> 
> 1. (style) spaces required around that '=' (ctx:VxV) - for a text in the help string,
> 2. FYI: Errors found in lines not modified in the patch - some pylint stuff outside the scope of the patch

Right, it doesn't look to be part of the PR but the workflow steps are still required and are distinct from the tests displayed in the blue Ocean view. @mjmac can this failure just be ignored?

**Comment by johannlombardi** [comment]
SPDK upgrade is landed. Please rebase.

File: src/control/server/mgmt_pool.go [hunk 17]
```diff
@@ -124,41 +125,57 @@ func (svc *mgmtSvc) calculateCreateStorage(req *mgmtpb.PoolCreateReq) error {
...
+	} else if len(req.GetTierratio()) == 1 {
+		req.Tierratio = append(req.Tierratio, 0)
 	}
 
 	storagePerRank := func(total uint64) uint64 {
 		return total / uint64(len(req.GetRanks()))
 	}
 
+	if len(req.Tierbytes) == 0 {
+		req.Tierbytes = make([]uint64, len(req.Tierratio))
```
Comment (tanabarr) [inline src/control/server/mgmt_pool.go:141]: why are these not named "req.TierBytes" and "req.TierRatio" ?

**Review by tanabarr — CHANGES_REQUESTED** [review]
Still some more work to do, happy to take on some of the changes if you don't mind me committing directly to the PR. Also:
- Does more forwarding code need removing from the provider?
- Some delicate formatting related code has been changed significantly in instance_storage{,_rpc}.go, @mjmac are you satisfied this has been validated with manual testing to cover edge cases that are not covered by system tests?

File: src/control/cmd/daos_firmware/handler.go [hunk 18]
```diff
@@ -92,10 +93,33 @@ type bdevHandler struct {
 
 func (h *bdevHandler) setupProvider(log logging.Logger) {
 	if h.bdevProvider == nil {
-		h.bdevProvider = bdev.DefaultProvider(log).WithForwardingDisabled()
+		h.bdevProvider = bdev.DefaultProvider(log)
 	}
 }
 
+// nvmeQueryHandler handles a request to update the NVMe device firmware from a file.
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/daos_firmware/handler.go:100]: ok

File: src/control/cmd/daos_firmware/handler.go [hunk 19]
```diff
@@ -92,10 +93,33 @@ type bdevHandler struct {
...
+type nvmeQueryHandler struct {
+	bdevHandler
+}
+
+// Handle updates the NVMe device firmware and returns the result as a pbin.Response.
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/daos_firmware/handler.go:105]: ok

File: src/control/cmd/daos_server/storage_test.go [hunk 20]
```diff
@@ -128,11 +126,12 @@ func TestDaosServer_StoragePrepare(t *testing.T) {
...
 					log: log,
 				},
-				scs: server.NewStorageControlService(log, bdevProvider, scmProvider, nil),
+				scs: server.NewStorageControlService(log, nil),
 			}
 
 			gotErr := cmd.Execute(nil)
-			common.CmpErr(t, tc.expErr, gotErr)
+			_ = gotErr
+			/*common.CmpErr(t, tc.expErr, gotErr)
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/daos_server/storage_test.go:134]: ok

File: src/control/cmd/dmg/pool_test.go [hunk 21]
```diff
@@ -87,55 +84,57 @@ func TestPoolCommands(t *testing.T) {
 		},
 		{
 			"Create pool with incompatible arguments (auto nvme-size)",
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/dmg/pool_test.go:86]: Disagree, IMO still imprecise. It depends how you look at it. The 1st arg is --size, which implies auto sizing. My understanding of arg incompatibility here is the use of both auto and manual args - in this case for nvme.

File: src/control/cmd/dmg/pool_test.go [hunk 22]
```diff
@@ -87,55 +84,57 @@ func TestPoolCommands(t *testing.T) {
...
-			fmt.Sprintf("pool create --size %s --nvme-size %s", testScmSizeStr, testScmSizeStr),
+			fmt.Sprintf("pool create --size %s --nvme-size %s", testSizeStr, testSizeStr),
 			"",
 			errors.New("may not be mixed"),
 		},
 		{
 			"Create pool with incompatible arguments (auto scm-size)",
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/cmd/dmg/pool_test.go:92]: Similar case to nvme.

File: src/control/server/ctl_storage_test.go [hunk 23]
```diff
@@ -6,19 +6,7 @@
...
-
-	"github.com/daos-stack/daos/src/control/common"
-	"github.com/daos-stack/daos/src/control/logging"
-	"github.com/daos-stack/daos/src/control/server/config"
-	"github.com/daos-stack/daos/src/control/server/engine"
-	"github.com/daos-stack/daos/src/control/server/storage"
-	"github.com/daos-stack/daos/src/control/server/storage/bdev"
-)
-
+/* todo_tiering
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/server/ctl_storage_test.go:9]: Disagree - these are related to VMD, agreed to work on it after this patch is merged.

File: src/control/server/engine/config_test.go [hunk 24]
```diff
@@ -81,7 +81,7 @@ func TestConfig_MergeEnvVars(t *testing.T) {
 	}
 }
 
-func TestConfig_HasEnvVar(t *testing.T) {
+func TestConfigHasEnvVar(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/server/engine/config_test.go:84]: ok

File: src/control/server/mgmt_pool.go (hunk 17 above)
Comment (llasek-intc) [reply src/control/server/mgmt_pool.go:141]: This is to follow the naming of other existing fields in the struct, like: Usergroup, Numsvcreps, Totalbytes, Numranks...

File: src/control/server/ctl_storage_rpc.go [hunk 25]
```diff
@@ -350,6 +339,7 @@ func (c *ControlService) StorageFormat(ctx context.Context, req *ctlpb.StorageFo
 		}
 	}
 
+	var err error = nil
```
[the remark this answers is not in the document]
Comment (llasek-intc) [reply src/control/server/ctl_storage_rpc.go:342]: > Still some more work to do, happy to take on some of the changes if you don't mind me committing directly to the PR. Also:

Please go ahead.

File: src/control/server/storage/config.go [hunk 26]
```diff
@@ -30,71 +34,244 @@ const (
...
+	return c
+}
+
+// WithBdevFileSize sets the backing file size (used when BdevClass is malloc or file).
+func (c *TierConfig) WithBdevFileSize(size int) *TierConfig {
+	c.Bdev.FileSize = size
+	return c
+}
+
+type TierConfigs []*TierConfig
```
Comment (tanabarr) [inline src/control/server/storage/config.go:162]: nit: exported types need a comment otherwise linter will complain (and documentation will suffer)

File: src/control/server/storage/config.go [hunk 27]
```diff
@@ -30,71 +34,244 @@ const (
...
+
+func (sc *Config) Validate() error {
```
Comment (tanabarr) [inline src/control/server/storage/config.go:164]: ditto

File: src/control/server/storage/provider.go [hunk 28]
```diff
@@ -0,0 +1,373 @@
...
+	return &ScmMountPoint{
+		Class:      cfg.Class,
+		DeviceList: cfg.Scm.DeviceList,
+		Path:       cfg.Scm.MountPoint,
+		TotalBytes: total,
+		AvailBytes: avail,
+	}, nil
+}
+
+func (p *Provider) ScmIsMounted() (bool, error) {
```
Comment (tanabarr) [inline src/control/server/storage/provider.go:78]: nit: missing comment

File: src/control/server/storage/provider.go [hunk 29]
```diff
@@ -0,0 +1,373 @@
...
+	cfg, err := p.GetScmConfig()
+	if err != nil {
+		return false, err
+	}
+	return p.Sys.IsMounted(cfg.Scm.MountPoint)
+}
+
+func (p *Provider) MountScm() error {
```
Comment (tanabarr) [inline src/control/server/storage/provider.go:86]: nit: missing comment

File: src/engine/init.c [hunk 30]
```diff
@@ -845,14 +849,15 @@ parse(int argc, char **argv)
 		{ "xshelpernr",		required_argument,	NULL,	'x' },
 		{ "instance_idx",	required_argument,	NULL,	'I' },
 		{ "bypass_health_chk",	no_argument,		NULL,	'b' },
+		{ "storage_tiers", required_argument, NULL, 'T' },
```
Comment (tanabarr) [inline src/engine/init.c:852]: nit: column alignment

File: src/proto/mgmt/pool.proto [hunk 31]
```diff
@@ -28,20 +28,18 @@ message PoolCreateReq {
 	repeated uint32 faultDomains = 7; // Fault domain tree, minimal format
 	uint32 numsvcreps = 8; // desired number of pool service replicas
 	uint64 totalbytes = 9; // Total pool size in bytes (auto config)
-	double scmratio = 10; // Ratio of SCM:NVMe expressed as % (auto config)
+	repeated double tierratio = 10; // Ratio of storage tiers expressed as % of totalbytes (auto config)
```
Comment (tanabarr) [inline src/proto/mgmt/pool.proto:31]: I seen now that the pool create request parameter names need updating to tier_ratio style notation

File: utils/config/daos_server.yml [hunk 32]
```diff
@@ -355,48 +358,51 @@
...
+#    # Options are:
+#    # - "dcpm" for real SCM (preferred option), scm_size is ignored
+#    # - "ram" to emulate SCM with memory, scm_list is ignored
+#    # Immutable after reformat.
+#
+#    class: dcpm
+#
+#    # When class is set to dcpm, scm_list is the list of device paths for
+#    # PMem namespaces (currently only one per engine supported).
+#    scm_list: [/dev/pmem0]
```
Comment (tanabarr) [inline utils/config/daos_server.yml:381]: This might be slightly confusing given the second engine is being assigned the first pmem device, I think it's more intuitive to map first and second pmem devices/sockets to first and second engines as it makes the resource allocation/upi usage mitigation story more obvious.

File: utils/config/examples/daos_server_sockets.yml [hunk 33]
```diff
@@ -46,38 +46,41 @@ engines:
   # - CRT_CTX_SHARE_ADDR=1
   # - CRT_CTX_NUM=8
 
-  # Storage definitions
-  scm_mount: /mnt/daos0
+  storage:
+  -
+    # Storage definitions
```
Comment (tanabarr) [inline utils/config/examples/daos_server_sockets.yml:51]: In my mind class should be at the top of each storage tier section, the user has to know which parameters to fill out for each class and I think it's easier if the class is given first.

File: utils/docker/Dockerfile.centos.8 [hunk 34]
```diff
@@ -162,6 +162,7 @@ RUN dnf -y upgrade && \
         python3-Cython \
         python3-devel \
         python3-distro \
+        python3-pyelftools \
```
Comment (tanabarr) [inline utils/docker/Dockerfile.centos.8:165]: this will break workflow builds, it needs to be installed directly with pip (as it currently is on master)

File: src/control/server/mgmt_pool.go (hunk 17 above)
Comment (tanabarr) [reply src/control/server/mgmt_pool.go:141]: yes this is technical debt in the Proto definitions for some of the request structures like for pool create request, it should be addressed outside of this PR: https://jira.hpdd.intel.com/browse/DAOS-8014

File: src/control/server/ctl_storage_rpc.go [hunk 35]
```diff
@@ -61,58 +59,46 @@ func mapCtrlrs(ctrlrs storage.NvmeControllers) (map[string]*storage.NvmeControll
...
 
 	for _, srv := range instances {
-		nvmeDevs := c.instanceStorage[srv.Index()].Bdev.GetNvmeDevs()
-		if len(nvmeDevs) == 0 {
+		if !srv.storage.HasBlockDevices() {
 			continue
 		}
+		direct := !srv.isReady()
 
-		// only retrieve results for devices listed in server config
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:74]: I will verify

File: src/control/server/ctl_storage_rpc.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:342]: as most of these comments are just nits I will resolve

File: src/control/server/ctl_storage_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/ctl_storage_test.go:9]: I will add ticket references to the commented sections

File: src/proto/mgmt/pool.proto (hunk 31 above)
Comment (tanabarr) [reply src/proto/mgmt/pool.proto:31]: https://jira.hpdd.intel.com/browse/DAOS-8014

**Review by johannlombardi — APPROVED** [review]
